scene/2d/physics_body_2d.cpp

```diff
@@ -999,17 +999,15 @@ Vector2 KinematicBody2D::move_and_slide(const Vector2 &p_linear_velocity, const
 	Vector2 body_velocity_normal = body_velocity.normalized();
 	Vector2 up_direction = p_up_direction.normalized();
 
-	Vector2 current_floor_velocity = floor_velocity;
-	if (on_floor && on_floor_body.is_valid()) {
-		//this approach makes sure there is less delay between the actual body velocity and the one we saved
-		PhysicsDirectBodyState2D *bs = PhysicsServer2D::get_singleton()->body_get_direct_state(on_floor_body);
-		if (bs) {
-			current_floor_velocity = bs->get_linear_velocity();
-		}
```

Comment:
As I've been writing, Immediately after, the KinematicBody will check colliding with using `body_velocity` and `move_and_collide()`. In fact, no actual motion penetration occur.

![kbfix_05](https://user-images.githubusercontent.com/61938263/94462957-343e9a80-01f7-11eb-9ac6-420fa2dd399d.gif)

Also, 2D was made to be the same implementation as 3D. The fix is [#34839](https://github.com/godotengine/godot/pull/34839) for 3D. If you don't have that problem in 2D, we can use `current_floor_velocity`. I'm sorry for the implicit changes.

Reply:
As I mentioned at the beginning, a moderate solution is adding an option to not calculate `floor_velocity` in `move_and_slide()` so that users can add `floor_velocity` to `global_transform`. What do you think? @madmiraal